### PR TITLE
[Merged by Bors] - feat: analogue of iSup.prod for sets

### PR DESCRIPTION
**Pull request by grunweg**
-----------

As suggested [on Zulip](https://leanprover.zulipchat.com/#narrow/stream/217875-Is-there-code-for-X.3F/topic/iUnion.20is.20compatible.20with.20product.20of.20types/near/395079870).

Feedback on the lemma name is appreciated; `iUnion_prod` is already taken by a different lemma.

### CODE REVIEW

File: Mathlib/Data/Set/Lattice.lean [hunk 1]
```diff
@@ -1872,6 +1872,11 @@ theorem iUnion_prod {ι ι' α β} (s : ι → Set α) (t : ι' → Set β) :
   simp
 #align set.Union_prod Set.iUnion_prod
 
+/-- Analogue of `iSup.prod` for sets. -/
+lemma Set.iUnion_prod' {X Y Z : Type*} (f : X × Y → Set Z) :
+    ⋃ (x : X) (y : Y), (f ⟨x, y⟩) = ⋃ t : X × Y, (f t) := by
```
[the remark this answers is not in the document]
Comment (grunweg) [reply Mathlib/Data/Set/Lattice.lean]: I have reversed the statement (and renamed all variables to exactly match `iSup_prod`. The proof becomes on line longer, though.

File: Mathlib/Data/Set/Lattice.lean [hunk 2]
```diff
@@ -1872,6 +1872,11 @@ theorem iUnion_prod {ι ι' α β} (s : ι → Set α) (t : ι' → Set β) :
   simp
 #align set.Union_prod Set.iUnion_prod
 
+/-- Analogue of `iSup.prod` for sets. -/
+lemma Set.iUnion_prod' (f : β × γ → Set α) : ⋃ x : β × γ, f x = ⋃ (i : β) (j : γ), (f ⟨i, j⟩) := by
+  symm
+  simp only [iUnion, iSup_eq_iUnion, iSup_prod]
```
Comment (alreadydone) [inline Mathlib/Data/Set/Lattice.lean:1878]:
```suggestion
lemma Set.iUnion_prod' (f : β × γ → Set α) : ⋃ x : β × γ, f x = ⋃ (i : β) (j : γ), f (i, j) := by
  simp only [iUnion, iSup_eq_iUnion, iSup_prod]
```
Interestingly, `⋃ (i) (j)` also works but `⋃ (i j)` and `⋃ i j` don't. I recall they all work in Lean 3 ...

**Comment by grunweg** [comment]
(Guessing a topic in the sense of mathlib.)

File: Mathlib/Data/Set/Lattice.lean [hunk 3]
```diff
@@ -1872,6 +1872,10 @@ theorem iUnion_prod {ι ι' α β} (s : ι → Set α) (t : ι' → Set β) :
   simp
 #align set.Union_prod Set.iUnion_prod
 
+/-- Analogue of `iSup.prod` for sets. -/
+lemma iUnion_prod' (f : β × γ → Set α) : ⋃ x : β × γ, f x = ⋃ (i : β) (j : γ), f (i, j) := by
+  simp only [iUnion, iSup_eq_iUnion, iSup_prod]
```
Comment (ADedecker) [inline Mathlib/Data/Set/Lattice.lean:1877]:
```suggestion
lemma iUnion_prod' (f : β × γ → Set α) : ⋃ x : β × γ, f x = ⋃ (i : β) (j : γ), f (i, j) :=
  iSup_prod
```

File: Mathlib/Data/Set/Lattice.lean [hunk 4]
```diff
@@ -1872,6 +1872,10 @@ theorem iUnion_prod {ι ι' α β} (s : ι → Set α) (t : ι' → Set β) :
   simp
 #align set.Union_prod Set.iUnion_prod
 
+/-- Analogue of `iSup.prod` for sets. -/
```
Comment (ADedecker) [inline Mathlib/Data/Set/Lattice.lean:1875]:
```suggestion
/-- Analogue of `iSup_prod` for sets. -/
```

**Review by ADedecker — APPROVED** [review]
Thanks!

bors d+

**Comment by grunweg** [comment]
Thank you for the fast review.

**Pull request closed without merging**